File: ios/Classes/Bridge/ManagerBuilderMethodHandler.swift
//
//  ManagerBuilderMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let channelName = "cleveradssolutions/manager_builder"

class ManagerBuilderMethodHandler: MethodHandler {
    private let consentFlowMethodHandler: ConsentFlowMethodHandler
    private let mediationManagerMethodHandler: MediationManagerMethodHandler

    private var builderField: CleverAdsSolutions.CASManagerBuilder?
    private var builder: CleverAdsSolutions.CASManagerBuilder {
        if let builder = builderField { return builder }
        let instance = CAS.buildManager()
        builderField = instance
        return instance
    }

    init(
        with registrar: FlutterPluginRegistrar,
        _ consentFlowMethodHandler: ConsentFlowMethodHandler,
        _ mediationManagerMethodHandler: MediationManagerMethodHandler
    ) {
        self.consentFlowMethodHandler = consentFlowMethodHandler
        self.mediationManagerMethodHandler = mediationManagerMethodHandler
        super.init(with: registrar, on: channelName)
    }

    override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "withTestAdMode": withTestAdMode(call, result)
        case "withAdTypes": withAdTypes(call, result)
        case "withUserId": withUserId(call, result)
        case "withConsentFlow": withConsentFlow(call, result)
        case "withMediationExtras": withMediationExtras(call, result)
        case "withFramework": withFramework(call, result)
        case "build": build(call, result)
        default: super.onMethodCall(call, result)
        }
    }

    private func withTestAdMode(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("isEnabled", result) { isEnabled in
            builder.withTestAdMode(isEnabled)
        }
    }

    private func withAdTypes(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("adTypes", result) { (adTypes: Int) in
            builder.withAdFlags(CASTypeFlags(rawValue: UInt(adTypes)))
        }
    }

    private func withUserId(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("userId", result) { userId in
            builder.withUserID(userId)
        }
    }

    private func withConsentFlow(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if let id: String = call.getArgAndCheckNil("id", result) {
            if let consentFlow = consentFlowMethodHandler[id] {
                builder.withConsentFlow(consentFlow)
                result(nil)
            } else {
                result(call.errorFieldNil("consentFlow"))
            }
        }
    }

    private func withMediationExtras(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("key", "value", result) { key, value in
            builder.withMediationExtras(value, forKey: key)
        }
    }

    private func withFramework(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("pluginVersion", result) { pluginVersion in
            builder.withFramework("Flutter", version: pluginVersion)
        }
    }

    private func build(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("id", result) { id in
            let manager = builder
                .withCompletionHandler({ [weak self] initialConfig in
                    self?.invokeMethod("onCASInitialized", [
                        "error": initialConfig.error as Any?,
                        "countryCode": initialConfig.countryCode,
                        "isConsentRequired": initialConfig.isConsentRequired,
                        "testMode": initialConfig.manager.isDemoAdMode,
                    ])
                })
                .create(withCasId: id)

            mediationManagerMethodHandler.setManager(manager)

            builderField = nil
        }
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/manager/ManagerBuilderMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge.manager

import com.cleveradssolutions.plugin.flutter.CASFlutterContext
import com.cleveradssolutions.plugin.flutter.bridge.ConsentFlowMethodHandler
import com.cleveradssolutions.plugin.flutter.CASChannel
import com.cleveradssolutions.plugin.flutter.errorFieldNull
import com.cleveradssolutions.plugin.flutter.getArgAndCheckNull
import com.cleveradssolutions.plugin.flutter.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.success
import com.cleversolutions.ads.android.CAS
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/manager_builder"

class ManagerBuilderMethodHandler(
    binding: FlutterPluginBinding,
    private val contextService: CASFlutterContext,
    private val consentFlowMethodHandler: ConsentFlowMethodHandler,
    private val mediationManagerMethodHandler: MediationManagerMethodHandler
) : CASChannel(binding, CHANNEL_NAME) {

    @Volatile
    private var builderField: CAS.ManagerBuilder? = null
    private val builder: CAS.ManagerBuilder
        get() {
            builderField?.let { return it }
            val instance = CAS.buildManager()
            builderField = instance
            return instance
        }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "withTestAdMode" -> withTestAdMode(call, result)
            "withAdTypes" -> withAdTypes(call, result)
            "withUserId" -> withUserId(call, result)
            "withConsentFlow" -> withConsentFlow(call, result)
            "withMediationExtras" -> withMediationExtras(call, result)
            "withFramework" -> withFramework(call, result)
            "build" -> build(call, result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun withTestAdMode(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            builder.withTestAdMode(it)
        }
    }

    private fun withAdTypes(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("adTypes", result) {
            builder.withEnabledAdTypes(it)
        }
    }

    private fun withUserId(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("userId", result) {
            builder.withUserID(it)
        }
    }

    private fun withConsentFlow(call: MethodCall, result: MethodChannel.Result) {
        val id = call.getArgAndCheckNull<String>("id", result) ?: return
        val consentFlow = consentFlowMethodHandler[id]
            ?: return result.errorFieldNull(call, "consentFlow")

        builder.withConsentFlow(consentFlow)

        result.success()
    }

    private fun withMediationExtras(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String, String>("key", "value", result) { key, value ->
            builder.withMediationExtras(key, value)
        }
    }

    private fun withFramework(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("pluginVersion", result) {
            builder.withFramework("Flutter", it)
        }
    }

    private fun build(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("id", result) {
            val manager = builder
                .withCasId(it)
                .withCompletionListener { config ->
                    invokeMethod(
                        "onCASInitialized",
                        hashMapOf(
                            "error" to config.error,
                            "countryCode" to config.countryCode,
                            "isConsentRequired" to config.isConsentRequired,
                            "testMode" to config.manager.isDemoAdMode
                        )
                    )
                }
                .build(contextService.activity ?: contextService.application)

            mediationManagerMethodHandler.setManager(manager)

            builderField = null
        }
    }
}